src/Shared/PackageManagers/ProjectManagerFactory.cs

```diff
@@ -0,0 +1,128 @@
+                },
+                {
+                    nameof(NPMProjectManager), destinationDirectory =>
+                        new NPMProjectManager(httpClientFactory, destinationDirectory)
+                },
+                {
+                    nameof(NuGetProjectManager), destinationDirectory =>
+                        new NuGetProjectManager(httpClientFactory, destinationDirectory, nugetPackageActions ?? new NuGetPackageActions()) // Add the NuGetPackageActions to the NuGetProjectManager.
+                },
+                {
+                    nameof(PyPIProjectManager), destinationDirectory =>
+                        new PyPIProjectManager(httpClientFactory, destinationDirectory)
+                },
+                {
+                    nameof(UbuntuProjectManager), destinationDirectory =>
+                        new UbuntuProjectManager(httpClientFactory, destinationDirectory)
+                },
+                {
+                    nameof(URLProjectManager), destinationDirectory =>
+                        new URLProjectManager(httpClientFactory, destinationDirectory)
+                },
+                {
+                    nameof(VSMProjectManager), destinationDirectory =>
+                        new VSMProjectManager(httpClientFactory, destinationDirectory)
+                },
+            };
+        }
+
+        /// <summary>
+        /// Creates an appropriate project manager for a package given its PackageURL.
+        /// </summary>
+        /// <param name="purl">The <see cref="PackageURL"/> for the package to create the project manager for.</param>
+        /// <param name="destinationDirectory">The new destination directory, if provided.</param>
+        /// <returns>The implementation of <see cref="BaseProjectManager"/> for this <paramref name="purl"/>'s type.</returns>
+        public BaseProjectManager? CreateProjectManager(PackageURL purl, string destinationDirectory = ".")
+        {
+            ConstructProjectManager? projectManager = _projectManagers.GetValueOrDefault($"{purl.Type}ProjectManager");
+
+            return projectManager?.Invoke(destinationDirectory);
+        }
+
+        /// <summary>
+        /// Add static method to just get a <see cref="BaseProjectManager"/> implementation.
+        /// </summary>
+        /// <param name="packageUrl">The <see cref="PackageURL"/> to get a project manager for.</param>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to optionally add.</param>
+        /// <param name="nugetPackageActions">The <see cref="NuGetPackageActions"/> to use in the <see cref="NuGetProjectManager"/>.</param>
+        /// <returns>A new <see cref="BaseProjectManager"/> implementation.</returns>
+        public static BaseProjectManager? GetProjectManager(PackageURL packageUrl, IHttpClientFactory? httpClientFactory = null, NuGetPackageActions? nugetPackageActions = null)
```

Comment:
```suggestion
        public static BaseProjectManager? GetProjectManager(PackageURL packageUrl, IHttpClientFactory? httpClientFactory = null, IManagerPackageActions? managerPackageActions = null)
```

Reply:
Also don't think this is applicable base on my previous comment: https://github.com/microsoft/OSSGadget/pull/318#discussion_r841916574

Reply:
See newer comment. This is not a scalable approach and must be changed.